### PR TITLE
Constr.of_kind preserve more sharing

### PR DESCRIPTION
**Pull request by SkySkimmer**
Also fix enforcement of the VM/native cast collapse which was
implemented in mkCast but missing in of_kind

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
Doesn't seem to do much but I'd rather try to be consistent with what we enforce for constr values so I propose we merge

**Comment by SkySkimmer** [comment]
@coqbot run full ci

File: kernel/constr.ml [hunk 1]
```diff
@@ -478,6 +335,158 @@ let decompose_app c =
...
+   - applicative terms have at least one argument and the
+     function is not itself an applicative term
+   - stacks of VM or native casts are collapsed
+   - small rels are shared
+   - small sorts are shared
+*)
+let of_kind = function
+| Rel n when 0 <= n && n < Array.length rels -> rels.(n)
+| App (f, [||]) -> f
+| App (f, a) when isApp f -> mkApp (f, a)
```
Comment (ppedrot) [inline kernel/constr.ml:370]: I really don't like when clauses, especially in a piece of code so low-level. Couldn't you just write
```ocaml
| App (App _ as f, a) -> mkApp (f, a)
```
or even just inlining mkApp in the branch?

Comment (SkySkimmer) [reply kernel/constr.ml]: ok

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot run full ci

**Comment by ppedrot** [comment]
@coqbot merge now

**Pull request merged**